patches/tModLoader/Terraria/ModLoader/ModBackgroundStyle.cs

```diff
@@ -43,44 +48,87 @@ protected override sealed void Register()
 	/// <summary>
 	/// Allows you to modify the transparency of all background styles that exist. In general, you should move the index equal to this style's slot closer to 1, and all other indexes closer to 0. The transitionSpeed parameter is what you should add/subtract to each element of the fades parameter. See the ExampleMod for an example.
 	/// </summary>
-	public abstract void ModifyFarFades(float[] fades, float transitionSpeed);
+	public abstract void ModifyStyleFade(float[] fades, float transitionSpeed);
 
 	/// <summary>
-	/// Allows you to determine which texture is drawn in the very back of the background. BackgroundTextureLoader.GetBackgroundSlot may be useful here, as well as for the other texture-choosing hooks.
+	/// Gives you complete freedom over how the background is drawn. Return true for ChooseCloseTexture, ChooseCloseMidTexture, ChooseCloseFarTexture, ChooseMiddleTexture and ChooseFarTexture to have an effect; return false to disable tModLoader's own code for drawing the close background.
 	/// </summary>
-	public virtual int ChooseFarTexture()
+	public virtual bool PreDrawBackground(BackgroundDistance distance, SpriteBatch spriteBatch, double backgroundTopMagicNumber, float bgGlobalScaleMultiplier, int pushBGTopHack)
+	{
+		return true;
+	}
+
+	/// <summary>
+	/// Allows you to determine which texture is drawn in the very back of the background. This also lets you modify the scale, parallax, Y value and Looping Width.
+	/// </summary>
+	/// <param name="scale">The scale.</param>
+	/// <param name="parallax">The parallax value.</param>
+	/// <param name="topY">The Top Y level that the BG is drawn at</param>
+	/// <param name="loopWidth">The looping width that the texture loops at. Defaults to (Main.screenWidth / Main.bgWidthScaled + 2)</param>
+	/// <returns></returns>
+	public virtual int ChooseFarTexture(ref float scale, ref double parallax, ref int topY, ref int loopWidth)
 	{
 		return -1;
 	}
 
 	/// <summary>
-	/// Allows you to determine which texture is drawn in the middle of the background.
+	/// Allows you to determine which texture is drawn in the middle of the background. This also lets you modify the scale, parallax, Y value and Looping Width.
 	/// </summary>
-	public virtual int ChooseMiddleTexture()
+	/// <param name="scale">The scale.</param>
+	/// <param name="parallax">The parallax value.</param>
+	/// <param name="topY">The Top Y level that the BG is drawn at</param>
+	/// <param name="loopWidth">The looping width that the texture loops at. Defaults to (Main.screenWidth / Main.bgWidthScaled + 2)</param>
+	/// <returns></returns>
+	public virtual int ChooseMiddleTexture(ref float scale, ref double parallax, ref int topY, ref int loopWidth)
```

Comment:
I have a feeling that someone will want to expand these overloads again in the future, so I'd prefer if this were a single-parameter overload that uses a special struct, `BackgroundTextureParameters`?.

Reply:
Boss bars have a `BossBarDrawParams` struct, so `BackgroundTextureParams`/`BackgroundLayerParams` for consistency.